pkg/controller/container-runtime-config/helpers.go

```diff
@@ -753,3 +783,214 @@ func convertICSPToIDMS(icsp *apioperatorsv1alpha1.ImageContentSourcePolicy) *api
+	if len(clusterScopePolicies) == 0 && len(scopeNamespacePolicies) == 0 {
+		return nil, nil
+	}
+
+	dockerConfig := make(map[string]DockerConfig)
+	sigstoreAttachment := DockerConfig{
+		UseSigstoreAttachments: true,
+	}
+	for scope := range clusterScopePolicies {
+		dockerConfig[scope] = sigstoreAttachment
+	}
+	for scope := range scopeNamespacePolicies {
+		dockerConfig[scope] = sigstoreAttachment
+	}
+
+	registriesConfig := &RegistriesConfig{}
+	registriesConfig.Docker = dockerConfig
+	data, err := yaml.Marshal(registriesConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling regisres configuration for sigstore (cluster)imagepolicies: %w", err)
+	}
+	return data, nil
+}
+
+// updatePolicyObjWithScopeRequirements adds signature requirements to docker transport of the non nil policyObj
+func updatePolicyObjWithScopeRequirements(policyObj *signature.Policy, requirements map[string]signature.PolicyRequirements) {
+	if policyObj.Transports == nil {
+		policyObj.Transports = make(map[string]signature.PolicyTransportScopes)
+	}
+	if _, ok := policyObj.Transports["docker"]; !ok {
+		policyObj.Transports["docker"] = make(map[string]signature.PolicyRequirements)
+	}
+	for scope, requirement := range requirements {
+		policyObj.Transports["docker"][scope] = append(policyObj.Transports["docker"][scope], requirement...)
+	}
+}
+
+// namespacePolicyRequirements collects imagepolicies into map[namespace]map[scope]policyRequirements
+func namespaceScopePolicyRequirements(scopeNamespacePolicies map[string]map[string]signature.PolicyRequirements) map[string]map[string]signature.PolicyRequirements {
+	policies := make(map[string]map[string]signature.PolicyRequirements)
+	for scope, namespacePolicies := range scopeNamespacePolicies {
+		for namespace, requirements := range namespacePolicies {
+			if policies[namespace] == nil {
+				policies[namespace] = make(map[string]signature.PolicyRequirements)
+			}
+			policies[namespace][scope] = append(policies[namespace][scope], requirements...)
+		}
+	}
+	return policies
+}
+
+func generateNamespacedPolicyJSON(namespaceJSONBase []byte, policies map[string]map[string]signature.PolicyRequirements) (map[string][]byte, error) {
+	namespaceJSONs := make(map[string][]byte)
+	decoder := json.NewDecoder(bytes.NewBuffer(namespaceJSONBase))
+
+	for namespace, requirements := range policies {
+
+		policyObj := &signature.Policy{}
+
+		err := decoder.Decode(policyObj)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding policy json for namespaced policies: %w", err)
+		}
+		updatePolicyObjWithScopeRequirements(policyObj, requirements)
+		data, err := json.Marshal(policyObj)
+		if err != nil {
+			return nil, fmt.Errorf("error marshalling policy json for namespaced policies: %w", err)
+		}
+		namespaceJSONs[namespace] = data
+	}
+	return namespaceJSONs, nil
+}
+
+func imagePolicyConfigFileList(namespaceJSONs map[string][]byte, sigstoreRegistriesConfigYaml []byte) []generatedConfigFile {
```

Comment:
CRI-O has to this point no ability to correctly implement per-namespace policies (that has stalled in https://github.com/cri-o/cri-o/pull/7046 , and my further work on that is tracked in https://issues.redhat.com/browse/RUN-1980 , so isn’t this rather premature?

Of course the code proposed in this PR could be reviewed concurrently, but I don’t think it makes sense to merge it.

Reply:
@mtrmac thank you for the clarification! Do you have plans to continue the CRI-O work (RUN-1980) in the near future?

Reply:
I’m definitely interested seeing this through to completion, after all the effort, but I don’t think I can spare the time in the next ~month at least.

Reply:
Note https://issues.redhat.com/browse/OTA-1170 ; AFAICS that would benefit from MCO implementing `ClusterImagePolicy` .

Reply:
I agree that sounds like a plan. If we can deliver the namespaced policies in CRI-O v1.30 (late April) then we can pull those into the MCO implementation afterwards.

Focusing on the cluster policy sounds totally reasonable for now. :+1:

Reply:
@mtrmac @saschagrunert PR https://github.com/openshift/machine-config-operator/pull/4160 implementing ClusterImagePolicy is ready for review.